### DISCUSSION TITLE
How to choose plugin hooks about analyze source code dependencies

### DISCUSSION THREAD

**Discussion by yangguansen** [webpack/webpack#15493] [Q&A] [answered]
I have a question, please help me. I had ask questions on StackOverflow(https://stackoverflow.com/questions/71389265/webpackhow-to-choose-plugin-hooks-about-analyze-source-code-dependencies) and Gitter, but nobody answer me.

Question:
I want to make a webpack plugin about analyze source code dependencies,it can generate depend tree to img from entry file. How to choose hooks about analyze source code dependencies? There are so much hooks in documents.

I tried code like this
```
const pluginName = 'TestPlugin';
let currentFile;
const array = [];
let node = {};

class TestPlugin {
  apply(compiler) {
    compiler.hooks.normalModuleFactory.tap(pluginName, (factory) => {
      factory.hooks.parser
        .for('javascript/auto')
        .tap(pluginName, (parser, options) => {
          parser.hooks.import.tap(pluginName, (statement, source) => {
            if (parser.state.current.resource.includes('node_modules')) {
              return;
            }

            const { rawRequest: fileName, resource: filePath } = parser.state.current;
            if (filePath !== currentFile) {
              currentFile = filePath;
              if (JSON.stringify(node) !== '{}') {
                array.push(node);
              }
              node = {
                name: fileName,
                path: filePath,
                children: []
              };
            }
            node.children.push({ name: source, value: source });
          });
        });

    // end
    compiler.hooks.compilation.tap(pluginName, (compilation) => {
      compilation.hooks.finishModules.tap(pluginName, (modules) => {
        debugger;
        console.log(modules);
      })
    })
  }
}

module.exports = TestPlugin;

```
Now I have the flatter array, But I could't take the array to tree. Can anybody help me?

**Comment by vankop** [comment, marked as answer]
if you need some work with ast you need `parser` hooks.

> Now I have the flatter array, But I could't take the array to tree. Can anybody help me?

you should use module graph for that. something like
```js
for (const [ name, { dependencies: deps}] of compilation.entries) {
  const dep = deps[deps.length - 1];
  const module = compilation.moduleGraph.getModule(dep);
  const connections = compilation.moduleGraph.getOutgoingConnections(module);
  // visit modules in connections
}
```

**Reply by yangguansen** [reply]
Yeah, I had make the tree succeed by flatten array, thanks.

**Reply by yangguansen** [reply]
And I have a other question, I could't get import trigger in `const component = lazy(() => import('code path'))` by parser.hooks.import.tap(). How could I get lazy load import hooks?

**Reply by vankop** [reply]
not sure what you mean.. all `import()` should be catched by `import` hook.

**Reply by yangguansen** [reply]
In react program, I define a component by `const Home = lazy(() => import(/* webpackChunkName: "Home" */'@src/routes/index'));`,
and I catch hook in webpack plugin 
```
  compiler.hooks.normalModuleFactory.tap(this.pluginName, (factory) => {
      factory.hooks.parser
        .for('javascript/auto')
        .tap(this.pluginName, (parser, options) => {

          parser.hooks.import.tap(this.pluginName, (statement, source) => {
            // It can only catch sync import statement, not catch lazy load import
        

            
          });
        });
    });
```

**Reply by vankop** [reply]
you should use `javascript/auto` and `javascript/esm`

**Reply by vankop** [reply]
also take a look how to get parser correctly https://github.com/webpack/webpack/blob/main/lib/dependencies/ImportPlugin.js

**Reply by yangguansen** [reply]
emm..
The code `const Home = lazy(() => import(/* webpackChunkName: "Home" */'@src/routes/index'));` is only a common file named `index.js` or `index.jsx`. I think `javascript/auto` is enough, and the file isn't a ES module, why need `javascript/esm`?  
Could you explain some detail ? please, Thanks

**Reply by yangguansen** [reply]
Then I'm sure `javascript/auto` `javascript/esm` `javascript/dynamic` is useless for react lazy route `import()`, all hooks is 
```
compiler.hooks.normalModuleFactory.tap(this.pluginName, (factory) => {
      factory.hooks.parser
        .for('javascript/auto')
        .tap(this.pluginName, (parser, options) => {
          parser.hooks.import.tap(this.pluginName, (statement, source) => {
            
            if (parser.state.current.resource.includes('node_modules')) {
              return;
            }
            // can't catch `import()` statement in my react route file `route.js`
            
          });
        });
    });
```

**Reply by yangguansen** [reply]
And I check async module in route module, it's ImportDenpendenciedBlock 
![image](https://user-images.githubusercontent.com/24775952/157586416-27884ce3-0300-4062-a8a6-c92f3bcd0926.png)
in all my sync module, the module's blocks array is empty,
like this:
![image](https://user-images.githubusercontent.com/24775952/157586507-e6dcb00e-2a34-48ee-be80-800f29f3e50a.png)
 so what the difference between import async module and import sync module?

**Reply by yangguansen** [reply]
And  I check the compilation source code, there isn't moduleGraph method...

**Reply by yangguansen** [reply]
Oh man, getOutgoingConnections is webpack 5, but I use webpack 4.

**Reply by yangguansen** [reply]
Maybe I should find some call methods  in lib/ModuleGraph.js

**Reply by yangguansen** [reply]
Oh bro, module doesn't has ModuleGraph class, right? I'm crazy
